Libraries/acdriver/Sources/CompileActionImageSet.cpp

```diff
@@ -0,0 +1,490 @@
+    }
+    png_read_image(png_struct_ptr, row_pointers);
+
+    /* clean up */
+    png_read_end (png_struct_ptr, info_struct_ptr);
+    png_destroy_read_struct (&png_struct_ptr, &info_struct_ptr, (png_infopp) NULL);
+    free(row_pointers);
+    fclose(fp);
+
+    if (width_out) {
+        *width_out = width;
+    }
+    if (height_out) {
+        *height_out = height;
+    }
+    if (channels_out) {
+        *channels_out = channels;
+    }
+    return true;
+}
+
+static uint16_t
+GenerateIdentifier(void) {
+	static uint16_t last = 0;
+	last = last + 1;
+	return last;
+}
+
+static bool ends_with(const std::string &str, const std::set<std::string> &endings)
+{
+    std::string lower(str.size(), '\0');
+    std::transform(str.begin(), str.end(), lower.begin(), ::tolower);
+
+    for (auto ending : endings) {
+        if (lower.size() >= ending.size() &&
+            lower.substr(lower.size()-ending.size(), lower.size()-1) == ending) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool
+CompileContents(ext::optional<car::Writer> &writer, std::string ns, std::shared_ptr<xcassets::Asset::Asset> const &parent, xcassets::Asset::ImageSet::Image const &image, Result *result)
+{
+    static std::map<std::string, uint16_t> idMap = {};
+    std::string name = FSUtil::GetBaseNameWithoutExtension(parent->path());
+
+    /* Skip any entry that is not attached to a file */
+    if (!image.fileName()) {
+        return true;
+    }
+
+    std::string filename = parent->path() + std::string("/") + *image.fileName();
+
+    /* An image without an idiom is considered unassigned */
+    if (!image.idiom()) {
+        return false;
+    }
+
+    /* scale defaults to 0 / all */
+    double scale = 0;
+    if (image.scale()) {
+        scale = image.scale()->value();
+    }
+
+    // TODO: We should filter by target-device / device-model / os-version
+    uint16_t idiom = car_attribute_identifier_idiom_value_universal;
+    switch(*image.idiom()) {
+        case xcassets::Slot::Idiom::Universal:
+            idiom = car_attribute_identifier_idiom_value_universal;
+            break;
+        case xcassets::Slot::Idiom::Phone:
+            idiom = car_attribute_identifier_idiom_value_phone;
+            break;
+        case xcassets::Slot::Idiom::Pad:
+            idiom = car_attribute_identifier_idiom_value_pad;
+            break;
+        case xcassets::Slot::Idiom::Desktop:
+            // TODO: desktop has no idiom value
+            break;
+        case xcassets::Slot::Idiom::TV:
+            idiom = car_attribute_identifier_idiom_value_tv;
+            break;
+        case xcassets::Slot::Idiom::Watch:
+            idiom = car_attribute_identifier_idiom_value_watch;
+            break;
+    }
+
+    std::vector<unsigned char> pixels;
+    size_t width = 0;
+    size_t height = 0;
+    size_t channels = 0;
+    car::Rendition::Data::Format format;
+
+    if (ends_with(filename, {std::string("png")})) {
```

Comment:
Yeah, I should have checked FSUtil. Maybe it should also do the tolower transform.